### ISSUE TITLE
Unexpected EnvoyFilter patch outcome for typed_extension_protocol_options

### ISSUE THREAD

**Issue by LithiumSR** [istio/istio#42424]
### Bug Description

Hello everyone,
I was tinkering with the EnvoyFilter API to merge a patch for an Envoy cluster and I got an unexpected outcome.
I have a cluster with the following upstream_http_protocol_options as starting situation:
```
      "typed_extension_protocol_options": {
       "envoy.extensions.upstreams.http.v3.HttpProtocolOptions": {
        "@type": "type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions",
        "upstream_http_protocol_options": {
         "auto_sni": true
        },
        "explicit_http_config": {
         "http2_protocol_options": {}
        }
       }
      },
```
I wanted to add a new option to http2_protocol_options so I wrote the following EnvoyFilter:
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: eric-nrf-service-mesh-sbi-traffic-auto-sni
spec:
  configPatches:
  - applyTo: CLUSTER
    match:
      cluster:
        portNumber: 443
        service: '*.example.org'
      context: SIDECAR_OUTBOUND
    patch:
      operation: MERGE
      value:
        typed_extension_protocol_options:
          envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
            '@type': type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
            explicit_http_config:
              http2_protocol_options:
                override_stream_error_on_invalid_http_message: false

```
Sadly, the outcome of patching completely removed the previous upstream http options (no more auto_sni=true), and I got the following outcome
```
explicit_http_config
"typed_extension_protocol_options": {
       "envoy.extensions.upstreams.http.v3.HttpProtocolOptions": {
        "@type": "type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions",
        "explicit_http_config": {
         "http2_protocol_options": {
          "override_stream_error_on_invalid_http_message": false
         }
        }
       }
      }
     },
```
Is this the expected behavior, did I wrote the patch wrongly?

Steps to reproduce (see [repro.zip](https://github.com/istio/istio/files/10216936/repro.zip) for yaml files):
1. Install Istio 1.16.1 with `istioctl install --set profile=demo --set values.pilot.env.ENABLE_AUTO_SNI=true`
2. Label ns with `kubectl label ns istio-system istio-injection=enabled`
3. Install sleep via `kubectl apply -f https://raw.githubusercontent.com/istio/istio/master/samples/sleep/sleep.yaml -n istio-system`
4. Deploy ServiceEntry and DR via `kubectl apply -f se_dr.yaml -n istio-system`
You can verify that the outbound|443||*.example.org cluster has auto_sni=true
5. Apply the EF with merge logic to patch the cluster to include override_stream_error_on_invalid_http_message in http2_protocol_options. `kubectl apply -f ef.yaml -n istio-system`
6. Verify that override_stream_error_on_invalid_http_message was added to outbound|443||*.example.org but auto_sni in 443 was unexpectedly removed.

### Version

```prose
client version: 1.16.1
control plane version: 1.16.1
data plane version: 1.16.1 (3 proxies)

Client Version: v1.19.2
Server Version: v1.25.2
```

### Additional Information

[bug-report.tar.gz](https://github.com/istio/istio/files/10216925/bug-report.tar.gz)

**Comment by zirain** [comment]
I canot reproduct this:
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: fortio-ext
  namespace: istio-system
spec:
  addresses:
  - 10.141.189.42
  hosts:
  - '*.example.org'
  location: MESH_EXTERNAL
  ports:
  - name: http-port-for-tls-origination
    number: 443
    protocol: HTTP2
  resolution: NONE
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: eric-nrf-service-mesh-sbi-traffic-auto-sni
  namespace: istio-system
spec:
  configPatches:
  - applyTo: CLUSTER
    match:
      cluster:
        portNumber: 443
        service: '*.example.org'
      context: SIDECAR_OUTBOUND
    patch:
      operation: MERGE
      value:
        typed_extension_protocol_options:
          envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
            '@type': type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
            explicit_http_config:
              http2_protocol_options:
                override_stream_error_on_invalid_http_message: false
```


![image](https://user-images.githubusercontent.com/4354057/207302849-a187049e-c6eb-479e-95e9-7b4ec6d81a3c.png)

**Comment by LithiumSR** [comment]
Hello @zirain 
Did you use ENABLE_AUTO_SNI=true when deploying istio controller?

**Comment by zirain** [comment]
> Hello @zirain Did you use ENABLE_AUTO_SNI=true when deploying istio controller?

IC, let me try again.

**Comment by ramaraochavali** [comment]
I think that seems expected behaviour.

**Comment by howardjohn** [comment]
If you look at the actual proto its `TypedExtensionProtocolOptions map[string]*any.Any`. Any is just a `bytes value = 2;`. There is no smart merging of `bytes`.

**Comment by LithiumSR** [comment]
Thank you @howardjohn for the info :)
I will close the ticket

**Comment by SpecialYang** [comment]
We have met this issue in production. We apply one EnvoyFilter to all cluster to enable preserving header case but override some h2 service settings result in 503. How can we do fine-granted merging?
